### ISSUE TITLE
[J3.7.x] Changes to addStyleSheetVersion, addStyleSheet, addScriptVersion, addScript, existing J3.6.x code is partly broken

### ISSUE THREAD

**Issue by ggppdk** [joomla/joomla-cms#12828]
### Steps to reproduce the issue
Try to use:
```php
JFactory::getDocument->addStyleSheetVersion('my.css', $version='3.1.1');
JFactory::getDocument->addStyleSheetVersion('my.js', $version='3.1.1');
```

### Expected result
CSS and JS files will be added to the page, together their version string

### Actual result
**J3.6.x**:  It works
**Staging**: Partly broken
- assets are added but without the version string
- you get PHP warning

Warning: Illegal string offset 'version' in deve/joomla-cms/libraries/joomla/document/document.php on line 550
Warning: Illegal string offset 'version' in devel/joomla-cms/libraries/joomla/document/document.php on line 741

### System information (as much as possible)
**J3.6.x signatures**
```php
public function addStyleSheetVersion($url, $version = null, $type = "text/css", $media = null, $attribs = array())
public function addScriptVersion($url, $version = null, $type = "text/javascript", $defer = false, $async = false)
```

**J3.7.x signatures**
```php
public function addStyleSheetVersion($url, $options = array(), $attribs = array())
public function addScriptVersion($url, $options = array(), $attribs = array())
```

At minimum the existing code of calling the method with the old way needs to work,
and there should be no PHP warnings

It should not be too difficult to detect the old way of calling the methods, and make them work

### Additional comments

_Also, in the rare case that someone **both** extends the JDocument, and also overrides the 2 methods, there will be some PHP warnings_

**Comment by ggppdk** [comment]
I looked at the code in the 
- i see that the code already tries to detect old way of calling

@andrepereiradasilva 

```php
if (!is_array($options) && !is_array($attribs))
```

should probably be: (**in all 4 methods**)
```php
if (!is_array($options) || !is_array($attribs))
```

**Comment by ggppdk** [comment]
Also for the case of B/C of someone extending JDocument and redeclaring the methods:
**addStyleSheet, addStyleSheetVersion
addScriptVersion, addScriptVersionVersion**

- i think the method parameters, need to go back **to the old default values** so that they are B/C with anyone extending the classes and overriding the methods

Notes:
- we can keep the new parameter names, as these are local to the method and make no B/C break
- **also no B/C break should be the removal of the extra parameters**, (as long as we have code to handle them inside the method, and i see that this code is there), thus any extending class override these method can still declare the old method parameters without problems

To summarize, $options and $attribs need their old defaults to stay compatible
addStyleSheet($url, $options = array(), $attribs = array())
addStyleSheetVersion($url, $options = array(), $attribs = array())
addScriptVersion($url, $options = array(), $attribs = array())
addScriptVersionVersion($url, $options = array(), $attribs = array())

**Comment by andrepereiradasilva** [comment]
@ggppdk i can try and solve the warning issues - it's obvious a bug.

Reggarding potential B/C issues with the change with method signatures in all 6 methods (they where changed in JHtml too), i asked to check B/C policy in the first PR that proposed this change in method signatures (https://github.com/joomla/joomla-cms/pull/11289).

If that are B/C problems in those PR they need all to be reverted.

For info, the sequence of PR that made this changes are:
- For JS: https://github.com/joomla/joomla-cms/pull/11289, https://github.com/joomla/joomla-cms/pull/12373, https://github.com/joomla/joomla-cms/pull/12429, https://github.com/joomla/joomla-cms/pull/11289 and https://github.com/joomla/joomla-cms/pull/12373.
- For CSS: https://github.com/joomla/joomla-cms/pull/12375 and https://github.com/joomla/joomla-cms/pull/12720, https://github.com/joomla/joomla-cms/pull/12781, https://github.com/joomla/joomla-cms/pull/12720 and https://github.com/joomla/joomla-cms/pull/12714.

Also all the merged PR that use this new PR method signatures would need to be corrected..
All all the js polyfill changes and all the PR that do inline js changes to use this new ie conditional option would needs to be changed/reverted.

(and probably more PR that i can't remember now)

So we need to be sure this is a B/C issue. PLT please advice.

**Comment by ggppdk** [comment]
Besides the small bug (wrong if check)

the only B/C problem that might be a problem (or it may not be, is the default values) of the methods
- but now that i rethink of it , default values are not a B/C break at all,

i was confused with declaration of type for method parameters
i will test now

**Comment by ggppdk** [comment]
@andrepereiradasilva 

I am sorry i was confused with type declaration for parameter,
i tested, no B/C break

Only the bug with PHP warning and the wrong IF statement needs to be: fixed
```php
if (!is_array($options) && !is_array($attribs))
```

**Comment by andrepereiradasilva** [comment]
np, will check that when i have time

**Comment by andrepereiradasilva** [comment]
@ggppdk i'm now looking with tiume to your PR description.

you are making this:
```php
JFactory::getDocument->addStyleSheetVersion('my.css', $version='3.1.1');
JFactory::getDocument->addStyleSheetVersion('my.js', $version='3.1.1');
```
when it should be this:
```php
JFactory::getDocument->addStyleSheetVersion('my.css', '3.1.1');
JFactory::getDocument->addStyleSheetVersion('my.js', '3.1.1');
```

**Comment by andrepereiradasilva** [comment]
@ggppdk plese teste/review https://github.com/joomla/joomla-cms/pull/12836

**Comment by jeckodevelopment** [comment]
Closing since we have a PR to test for it.
Thanks

**Comment by ggppdk** [comment]
@andrepereiradasilva 

Using **$version='3.1.1'**,

is irrelevant, **it is about readability and about avoid bugs**

- you are simpling declaring a variable while calling the method
- **it does not change the type of the passed parameter**, in both cases it is 'string'

I usually do not use this, when only having 2 or 3 parameters
but for more i do for 3+ parameters

- Only when the code is going to be executed "many" times, then it should be avoided

Otherwise the **performance difference due to the redudant variable declaration is non-measurable**,
and probably PHP compiler, will disgard the variable declaration it completely

Example of readability, and **avoiding mistakes**:
```php
$send_result = JFactory::getMailer()->sendMail(
	$_from = $data['mailfrom'], $_data['fromname'], $_recipient = array($email),
	$_emailSubject, $_emailBody,
	$_html_mode=true, $_cc=null, $_bcc=null, $_attachment=null,
	$_replyto='some@sommmmmm.org', $_replytoname=null
);
```

You could of course declare the variables above the method call, but this means a few extra lines

**Comment by andrepereiradasilva** [comment]
sure please test the PR

**Comment by ggppdk** [comment]
doing already
